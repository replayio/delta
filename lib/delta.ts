import { Project } from "./server/supabase/supabase";

export const getDeltaBranchUrl = (project: Project, branchName: string) =>
  `https://delta.replay.io/project/${project.short}/?branch=${branchName}`;
