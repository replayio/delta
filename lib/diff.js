const PNG = require("pngjs").PNG;
const pixelmatch = require("pixelmatch");

function diffImages(img1, img2) {
  try {
    const png1 = PNG.sync.read(img1);
    const png2 = PNG.sync.read(img2);
    const { width, height } = png1;
    const diff = new PNG({ width, height });

    const numDiffPixels = pixelmatch(
      png1.data,
      png2.data,
      diff.data,
      width,
      height,
      {
        threshold: 0.2,
      }
    );

    const diffPng = PNG.sync.write(diff);
    const changed = numDiffPixels > 0;
    console.log({ numDiffPixels });

    return { changed, numDiffPixels, diffPng };
  } catch (e) {
    console.log("error", e);
    return { changed: true };
  }
}

module.exports = { diffImages };
