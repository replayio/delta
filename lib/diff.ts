import { PNG } from "pngjs";
import pixelmatch from "pixelmatch";

export function diffImages(img1, img2) {
  try {
    const png1 = PNG.sync.read(img1);
    const png2 = PNG.sync.read(img2);
    const { width, height } = png1;
    const diff = new PNG({ width, height });

    const numDiffPixels = pixelmatch(
      png1.data,
      png2.data,
      diff.data,
      width,
      height,
      {
        threshold: 0.2,
      }
    );

    const diffPng = PNG.sync.write(diff);
    const changed = numDiffPixels > 0;
    console.log({ numDiffPixels });

    return { changed, numDiffPixels, diffPng };
  } catch (e) {
    console.log("error", e);
    return { changed: true };
  }
}

export function diffBase64Images(img1, img2) {
  if (!img1 || !img2) {
    return { changed: false };
  }
  return diffImages(Buffer.from(img1, "base64"), Buffer.from(img2, "base64"));
}
