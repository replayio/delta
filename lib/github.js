const fetch = require("node-fetch");
const { Octokit } = require("@octokit/rest");
const dotenv = require("dotenv");

dotenv.config({ path: "./.env.local" });
const HOST = process.env.HOST;

async function fetchToken() {
  let res;

  try {
    res = await fetch(`${HOST}/api/getToken`);

    if (res.status === 200) {
      const body = await res.json();
      return body;
    }

    const body = await res.text();
    console.log(res.status, body);
    return body;
  } catch (e) {
    console.error("error", e);
  }
}

async function createCheck(
  owner,
  repo,
  { head_sha, title, summary, conclusion, text, status }
) {
  const tokenRes = await fetchToken();
  const octokit = new Octokit({ auth: tokenRes.token });

  return octokit.request(`POST /repos/${owner}/${repo}/check-runs`, {
    owner: owner,
    repo: repo,
    name: "Visuals",
    head_sha,
    status,
    conclusion,
    details_url: "https://replay-visuals.vercel.app/?branch=visuals5",
    started_at: new Date().toISOString(),
    output: {
      title,
      summary,
      text,
    },
  });
}

async function updateCheck(
  owner,
  repo,
  checkRunId,
  { head_sha, status, conclusion, title, summary, text }
) {
  const tokenRes = await fetchToken();
  const octokit = new Octokit({ auth: tokenRes.token });

  return octokit.request(
    `PATCH /repos/${owner}/${repo}/check-runs/${checkRunId}`,
    {
      owner,
      repo,
      check_run_id: checkRunId,
      name: "Visuals",
      head_sha: head_sha,
      status,
      conclusion,
      completed_at: new Date().toISOString(),
      output: {
        title,
        summary,
        text,
      },
    }
  );
}

module.exports = {
  createCheck,
  updateCheck,
};
