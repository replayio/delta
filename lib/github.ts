import { App } from "@octokit/app";
import dotenv from "dotenv";

dotenv.config({ path: "./.env.local" });

type CheckArgs = {
  head_sha?: string;
  status?: string;
  conclusion?: string;
  title?: string;
  summary?: string;
  text?: string;
};

async function getOctokit() {
  const pem = process.env.PEM;
  const appId = 274973;
  const owner = "replayio";
  const repo = "devtools";

  const app = new App({ appId, privateKey: pem });

  // First we need to get the installation id for the repo
  const { data: installation } = await app.octokit.request(
    `GET /repos/${owner}/${repo}/installation`
  );

  // Then we can get an octokit instance for the installation
  const octokit = await app.getInstallationOctokit(installation.id);
  return octokit;
}

export async function createCheck(
  owner,
  repo,
  { head_sha, title, summary, conclusion, text, status }
) {
  const octokit = await getOctokit();

  const check = octokit.request(`POST /repos/${owner}/${repo}/check-runs`, {
    owner: owner,
    repo: repo,
    name: "Visuals",
    head_sha,
    status,
    conclusion,
    details_url: "https://replay-visuals.vercel.app/?branch=visuals5",
    started_at: new Date().toISOString(),
    output: {
      title,
      summary,
      text,
    },
  });

  return check;
}

export async function updateCheck(
  owner: string,
  repo: string,
  checkRunId: string,
  { head_sha, status, conclusion, title, summary, text }: CheckArgs
) {
  const octokit = await getOctokit();

  const checkArgs = {
    owner,
    repo,
    check_run_id: checkRunId,
    name: "Visuals",
    head_sha: head_sha,
    status,
    conclusion,
    completed_at: new Date().toISOString(),
    output: {
      title,
      summary,
      text,
    },
  };

  console.log("updateCheck", checkArgs);
  return octokit.request(
    `PATCH /repos/${owner}/${repo}/check-runs/${checkRunId}`,
    checkArgs
  );
}
