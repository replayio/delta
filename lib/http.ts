import { safeStringify } from "./server/json";

const HOST = process.env.HOST;

export async function post(path: string, body: any) {
  const res = await fetch(`${HOST}${path}`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: safeStringify(body),
  });

  if (res.status === 200) {
    return { status: 200, data: res.json() };
  }

  return { status: res.status, error: await res.text() };
}
