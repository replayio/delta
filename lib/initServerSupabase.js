import { createClient } from "@supabase/supabase-js";
import dotenv from "dotenv";
dotenv.config({ path: "./.env.local" });

export default function initSupabase() {
  return createClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL,
    process.env.SUPABASE_SERVICE_KEY
  );
}
