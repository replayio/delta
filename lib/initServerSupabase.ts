import { createClient } from "@supabase/supabase-js";
import { config } from "dotenv";

config({ path: "./.env.local" });

export default function initSupabase() {
  return createClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.SUPABASE_SERVICE_KEY!
  );
}
