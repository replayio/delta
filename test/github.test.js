import { createComment } from "../lib/github";
import { describe, it, expect } from "vitest";
import omit from "lodash/omit";

describe("github", () => {
  it("create comment", async () => {
    const { data } = await createComment("replayio", "devtools", 8432);

    expect(
      omit(data, [
        "created_at",
        "html_url",
        "id",
        "node_id",
        "updated_at",
        "url",
        "reactions",
      ])
    ).toMatchSnapshot();
  });
});
