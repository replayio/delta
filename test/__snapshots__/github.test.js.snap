// Vitest Snapshot v1

exports[`github > create comment 1`] = `
{
  "author_association": "NONE",
  "body": "...",
  "issue_url": "https://api.github.com/repos/replayio/devtools/issues/8432",
  "performed_via_github_app": null,
  "user": {
    "avatar_url": "https://avatars.githubusercontent.com/in/274973?v=4",
    "events_url": "https://api.github.com/users/replay-delta%5Bbot%5D/events{/privacy}",
    "followers_url": "https://api.github.com/users/replay-delta%5Bbot%5D/followers",
    "following_url": "https://api.github.com/users/replay-delta%5Bbot%5D/following{/other_user}",
    "gists_url": "https://api.github.com/users/replay-delta%5Bbot%5D/gists{/gist_id}",
    "gravatar_id": "",
    "html_url": "https://github.com/apps/replay-delta",
    "id": 121197601,
    "login": "replay-delta[bot]",
    "node_id": "BOT_kgDOBzlUIQ",
    "organizations_url": "https://api.github.com/users/replay-delta%5Bbot%5D/orgs",
    "received_events_url": "https://api.github.com/users/replay-delta%5Bbot%5D/received_events",
    "repos_url": "https://api.github.com/users/replay-delta%5Bbot%5D/repos",
    "site_admin": false,
    "starred_url": "https://api.github.com/users/replay-delta%5Bbot%5D/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/replay-delta%5Bbot%5D/subscriptions",
    "type": "Bot",
    "url": "https://api.github.com/users/replay-delta%5Bbot%5D",
  },
}
`;
